lib/src/main/kotlin/StatefulSolver.kt

```diff
@@ -1,105 +1,82 @@
+        // If the node is terminal, return it
+        if (mdp.isTerminal(node.state)) {
+            return node
         }
 
-        // Expansion
-        val expandedState = expandNode(bestState)
+        val exploredActions = node.exploredActions()
 
-        if (verbose) {
-            traceln("Expanding:")
-            displayNode(expandedState)
+        // This state has not been fully explored
+        if (node.validActions.size != exploredActions.size) {
+            return node
         }
 
-        // Simulation
-        val simulatedReward = simulateState(expandedState)
-
-        traceln("Simulated Reward: $simulatedReward")
+        // This state has been explored, select best action
+        var bestAction = exploredActions.first()
+        var bestActionScore : Double? = null
+
+        for (action in exploredActions) {
+            val childrenOfAction = node.getChildren(action)
+            val actionN = childrenOfAction.sumOf { c -> c.n }
+            val actionReward = childrenOfAction.sumOf { c -> c.reward }
+            val actionScore = calculateUCT(node.n, actionN, actionReward, explorationConstant)
+
+            if (bestActionScore == null || actionScore > bestActionScore) {
+                bestAction = action
+                bestActionScore = actionScore
+            }
+        }
 
-        // Update
-        updateNode(expandedState, simulatedReward)
-    }
+        val newState = mdp.transition(node.state, bestAction)
 
-    private fun updateNode(stateNode: StateNode<TAction, TState>, simulatedReward: Double) {
-        var currentStateNode = stateNode
-        var currentReward = simulatedReward
+        val actionState = node.getChildren(bestAction).firstOrNull { s -> s.state == newState }
+        // New state reached by an explored action
+                ?: return createNode(node, bestAction, newState)
 
-        while (true)
-        {
-            currentStateNode.maxReward = max(currentReward, currentStateNode.maxReward)
-            currentStateNode.reward += currentReward
-            currentStateNode.n++
 
-            val parentActionNode = currentStateNode.parentAction() ?: break
-            parentActionNode.reward += currentReward
-            parentActionNode.n++
+        // Existing state reached by an explored action
+        return selectNode(actionState)
+    }
 
-            currentStateNode = parentActionNode.parentState()
-            currentReward *= rewardDiscountFactor
+    override fun expandNode(node: StateNode<TState, TAction>): StateNode<TState, TAction> {
+        // If the node is terminal, return it
+        if (node.isTerminal) {
+            return node
         }
+
+        // Expand an unexplored action
+        val unexploredActions = node.validActions.minus(node.getChildren().map { c -> c.inducingAction }).distinct()
+        val actionTaken = unexploredActions.random() ?: throw Exception("No unexplored actions available")
```

Comment:
For every elvis operator with an Exception, note that the current Unit tests do not pick this up when the code fails. Could we do something different, or edit the Unit tests in the app accordingly?

Reply:
Will close if cannot reproduce bug.